In [1]:
print("--- Demonstrating Modules ---")
import math
print(f"The value of pi from the math module is: {math.pi}")

# Note: To create your own module, you would typically save code in a .py file (e.g., mymodule.py)
# and then import it like: `import mymodule` or `from mymodule import my_function`

print("\n--- Demonstrating Functions ---")
# Define a simple function
def greet(name):
    """This function greets the person passed in as an argument."""
    return f"Hello, {name}!"

# Call the function
user_name = "Alice"
print(greet(user_name))

def add_numbers(a, b):
    return a + b

result = add_numbers(5, 3)
print(f"5 + 3 = {result}")

print("\n--- Demonstrating Try/Except ---")
# Example of handling a ValueError when converting input
try:
    user_input = input("Enter a whole number: ")
    number = int(user_input)
    print(f"You entered: {number}")
except ValueError:
    print(f"Error: '{user_input}' is not a valid whole number. Please try again.")

print("\n--- Demonstrating Input -> Processing -> Output Flow ---")
# A simple program combining all concepts
def calculate_square(num):
    """Calculates the square of a number."""
    return num * num

def main():
    print("Welcome to the Square Calculator!")
    while True:
        user_val = input("Please enter a number to square (or 'quit' to exit): ")
        if user_val.lower() == 'quit':
            print("Exiting Square Calculator. Goodbye!")
            break

        try:
            number_to_square = float(user_val) # Use float to allow decimal numbers
            squared_result = calculate_square(number_to_square)
            print(f"The square of {number_to_square} is: {squared_result}")
        except ValueError:
            print(f"Invalid input: '{user_val}'. Please enter a valid number.")

# Run the main program
main()

--- Demonstrating Modules ---
The value of pi from the math module is: 3.141592653589793

--- Demonstrating Functions ---
Hello, Alice!
5 + 3 = 8

--- Demonstrating Try/Except ---
Enter a whole number: 10
You entered: 10

--- Demonstrating Input -> Processing -> Output Flow ---
Welcome to the Square Calculator!
Please enter a number to square (or 'quit' to exit): 2
The square of 2.0 is: 4.0
Please enter a number to square (or 'quit' to exit): quit
Exiting Square Calculator. Goodbye!


In [5]:
import csv

data = [
    {"name": "Alice", "age": 30,"physics":90,"chemistry":98,"maths":97},
    {"name": "Bob", "age": 25,"physics":80,"chemistry":88,"maths":97},
    {"name": "Charlie", "age": 35,"physics":70,"chemistry":98,"maths":87}
]

with open('data.csv', 'w', newline='') as csvfile:
    fieldnames = ["name", "age","physics","chemistry","maths"]
    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
    writer.writeheader() # Add header row
    writer.writerows(data) # Write the data


In [6]:
with open('data.csv', 'r') as csvfile:
    reader = csv.DictReader(csvfile)
    for row in reader:
        print(row)

{'name': 'Alice', 'age': '30', 'physics': '90', 'chemistry': '98', 'maths': '97'}
{'name': 'Bob', 'age': '25', 'physics': '80', 'chemistry': '88', 'maths': '97'}
{'name': 'Charlie', 'age': '35', 'physics': '70', 'chemistry': '98', 'maths': '87'}


# Task
Analyze the student data from `data.csv` to calculate total marks, percentage, and assign a grade for each student, then display these results.

## Define Helper Functions

### Subtask:
Define Python functions to calculate total marks, percentage, and grade based on given scores. This will make the code modular and readable.


**Reasoning**:
I need to define three Python functions: `calculate_total_marks`, `calculate_percentage`, and `assign_grade` as per the instructions. This code block will implement these functions.



In [7]:
def calculate_total_marks(physics, chemistry, maths):
    """Calculates the total marks from individual subject scores."""
    return physics + chemistry + maths

def calculate_percentage(total_marks, max_total_marks):
    """Calculates the percentage based on total marks and maximum possible marks."""
    return (total_marks / max_total_marks) * 100

def assign_grade(percentage):
    """Assigns a letter grade based on the given percentage."""
    if percentage >= 90:
        return 'A'
    elif percentage >= 80:
        return 'B'
    elif percentage >= 70:
        return 'C'
    elif percentage >= 60:
        return 'D'
    else:
        return 'F'

print("Helper functions for calculating total marks, percentage, and grade have been defined.")

Helper functions for calculating total marks, percentage, and grade have been defined.


## Read Student Data

### Subtask:
Read the student data from 'data.csv' into a list of dictionaries, ensuring the marks are converted to numeric types for calculations.


**Reasoning**:
The subtask requires reading data from 'data.csv', converting specific fields to numeric types, and storing them in a list of dictionaries. This code block directly implements those steps.



In [8]:
students_data = []
MAX_SUBJECT_MARKS = 100 # Assuming each subject has a maximum of 100 marks
NUM_SUBJECTS = 3 # physics, chemistry, maths
MAX_TOTAL_MARKS = MAX_SUBJECT_MARKS * NUM_SUBJECTS

with open('data.csv', 'r') as csvfile:
    reader = csv.DictReader(csvfile)
    for row in reader:
        # Convert marks to integers
        row['physics'] = int(row['physics'])
        row['chemistry'] = int(row['chemistry'])
        row['maths'] = int(row['maths'])
        # Convert age to integer as well, as it might be used for analysis
        row['age'] = int(row['age'])
        students_data.append(row)

print("Student data loaded and marks converted to numeric types:")
for student in students_data:
    print(student)

Student data loaded and marks converted to numeric types:
{'name': 'Alice', 'age': 30, 'physics': 90, 'chemistry': 98, 'maths': 97}
{'name': 'Bob', 'age': 25, 'physics': 80, 'chemistry': 88, 'maths': 97}
{'name': 'Charlie', 'age': 35, 'physics': 70, 'chemistry': 98, 'maths': 87}


## Process and Calculate Results

### Subtask:
Iterate through each student's record, calculate their total marks, percentage, and assign a grade using the helper functions.


**Reasoning**:
To fulfill the subtask, I will iterate through the 'students_data' list and for each student, calculate their total marks, percentage, and assign a grade using the predefined helper functions. These new metrics will be added to each student's dictionary.



In [9]:
for student in students_data:
    total_marks = calculate_total_marks(student['physics'], student['chemistry'], student['maths'])
    percentage = calculate_percentage(total_marks, MAX_TOTAL_MARKS)
    grade = assign_grade(percentage)

    student['total_marks'] = total_marks
    student['percentage'] = percentage
    student['grade'] = grade

print("Calculated total marks, percentage, and grade for each student:")
for student in students_data:
    print(student)

Calculated total marks, percentage, and grade for each student:
{'name': 'Alice', 'age': 30, 'physics': 90, 'chemistry': 98, 'maths': 97, 'total_marks': 285, 'percentage': 95.0, 'grade': 'A'}
{'name': 'Bob', 'age': 25, 'physics': 80, 'chemistry': 88, 'maths': 97, 'total_marks': 265, 'percentage': 88.33333333333333, 'grade': 'B'}
{'name': 'Charlie', 'age': 35, 'physics': 70, 'chemistry': 98, 'maths': 87, 'total_marks': 255, 'percentage': 85.0, 'grade': 'B'}


## Display Results

### Subtask:
Print out the name, total marks, percentage, and grade for each student in a clear format.


## Summary:

### Data Analysis Key Findings

*   Helper functions were successfully defined for calculating total marks, percentage, and assigning grades, ensuring modularity.
*   Student data from `data.csv` was loaded, with 'physics', 'chemistry', 'maths', and 'age' fields correctly converted to numeric types.
*   Each student's record was processed to include calculated 'total\_marks', 'percentage', and 'grade'. For example, Alice achieved 285 total marks (95.0%), earning an 'A' grade, while Bob received 265 total marks (88.33%), resulting in a 'B' grade.

### Insights or Next Steps

*   The processed data is now ready for further analysis, such as identifying top-performing students, analyzing subject-wise performance, or exploring correlations between age and academic performance.
*   Consider implementing error handling for the CSV reading process and validation for input marks to ensure data integrity and robustness.
